jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2StreamEndPoint.java

```diff
@@ -611,6 +610,7 @@ private IOException ioFailure()
         private void succeed()
         {
             callback.succeeded();
+            stream.demand(1);
```

Comment:
Why?

Reply:
This change, plus the ones in `HTTP2ServerConnectionFactory` and `Stream` were related to a bug in the H2's handling of backward compatibility calls (`onDataDemanded` vs `onData`) but the details are long gone from my mind. I modified those methods after a lengthy discussion with @sbordet so maybe he could remember the details?
Anyways, if my memory is right, those could go into a separate PR.

Reply:
What tests fail if it is removed?
I think having multiple things doing demand is at worst a big race and at best a source of unwanted events that make back pressure not work.

If we are demanding on every successful onData entry, then how sure are we that we are really getting back pressure from the application via HttpInput?  Isn't this demand here going to negate the demand we do in HttpChannelOverHttp2?

Reply:
I've investigated this and asked @sbordet for his help, and here's my this `demand()` call is required:

Before these changes, `Content` used to be queued and demand used to be chained automatically, i.e.: each time a demand was served, another one was automatically registered without `HttpChannelOverHTTP2` or any other class having to explicitly call `demand(1)`.

Since we're not queuing anymore, `HttpChannelOverHTTP2` needs explicit control over when demand is called, so `HTTP2ServerConnectionFactory` now overrides `onDataDemanded()` instead of `onData()` and we have to manually call `demand(1)` for serving `HttpInput` requests (as is done in `HttpChannelOverHTTP2.needContent()` and `produceContent()`) but also for other uses that do not make use of `HttpChannelOverHTTP2` like this tunnelling EndPoint that is used by WebSockets.